src/ray/gcs/tables.cc

```diff
@@ -373,55 +373,138 @@ void ClientTable::RegisterClientRemovedCallback(const ClientTableCallback &callb
+      resource_createupdated_callback_(client_, entry.first, entry.second);
+    }
+  }
+}
+
+void ClientTable::RegisterResourceDeletedCallback(const ClientTableCallback &callback) {
+  resource_deleted_callback_ = callback;
+  // Call the callback for any clients that are cached.
+  for (const auto &entry : client_cache_) {
+    if (!entry.first.is_nil() && (entry.second.entry_type == EntryType::RES_DELETE)) {
+      resource_deleted_callback_(client_, entry.first, entry.second);
+    }
+  }
+}
+
 void ClientTable::HandleNotification(AsyncGcsClient *client,
                                      const ClientTableDataT &data) {
   ClientID client_id = ClientID::from_binary(data.client_id);
   // It's possible to get duplicate notifications from the client table, so
   // check whether this notification is new.
   auto entry = client_cache_.find(client_id);
-  bool is_new;
+  bool is_notif_new;
   if (entry == client_cache_.end()) {
     // If the entry is not in the cache, then the notification is new.
-    is_new = true;
+    is_notif_new = true;
   } else {
     // If the entry is in the cache, then the notification is new if the client
-    // was alive and is now dead.
-    bool was_inserted = entry->second.is_insertion;
-    bool is_deleted = !data.is_insertion;
-    is_new = (was_inserted && is_deleted);
+    // was alive and is now dead or resources have been updated.
+    bool was_not_deleted = (entry->second.entry_type != EntryType::DELETION);
+    bool is_deleted = (data.entry_type == EntryType::DELETION);
+    bool is_res_modified = ((data.entry_type == EntryType::RES_CREATEUPDATE) ||
+                            (data.entry_type == EntryType::RES_DELETE));
+    is_notif_new = (was_not_deleted && (is_deleted || is_res_modified));
     // Once a client with a given ID has been removed, it should never be added
     // again. If the entry was in the cache and the client was deleted, check
     // that this new notification is not an insertion.
-    if (!entry->second.is_insertion) {
-      RAY_CHECK(!data.is_insertion)
+    if (entry->second.entry_type == EntryType::DELETION) {
+      RAY_CHECK((data.entry_type == EntryType::DELETION))
           << "Notification for addition of a client that was already removed:"
           << client_id;
     }
   }
 
   // Add the notification to our cache. Notifications are idempotent.
-  client_cache_[client_id] = data;
+  // If it is a new client or a client removal, add as is
+  if ((data.entry_type == EntryType::INSERTION) ||
+      (data.entry_type == EntryType::DELETION)) {
+    RAY_LOG(DEBUG) << "[ClientTableNotification] ClientTable Insertion/Deletion "
+                      "notification for client id "
+                   << client_id << ". EntryType: " << int(data.entry_type)
+                   << ". Setting the client cache to data.";
+    client_cache_[client_id] = data;
+  }
+  // If the entry is resource creation or updation
```

Comment:
The dangled `else if` looks strange. This comment is not necessary because the code is self-explanatory.